File: CommonCode/SettingsStore.m
//
//  SettingsStore.m
//  SpectrogeddonOSX
//
//  Created by Tom York on 08/05/2014.
//  Copyright (c) 2014 Spectrogeddon. All rights reserved.
//

#import "SettingsStore.h"
#import "DisplaySettings.h"

static NSString* const StoreFileName = @"Spectrogeddon.settings";

@interface SettingsStore ()
@property (nonatomic,strong) DisplaySettings* storedDisplaySettings;
@end

@implementation SettingsStore

+ (NSString*)pathToStore
{
    NSArray* directories = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    return [directories.firstObject stringByAppendingPathComponent:StoreFileName];
}

- (id)init
{
    if((self = [super init]))
    {
        [self loadDisplaySettings];
        if(!_storedDisplaySettings)
        {
            _storedDisplaySettings = [[DisplaySettings alloc] init];
        }
    }
    return self;
}

- (DisplaySettings*)displaySettings
{
    return self.storedDisplaySettings;
}

- (void)applyUpdateToSettings:(DisplaySettings*(^)(DisplaySettings*))updateBlock
{
    NSParameterAssert(updateBlock);
    self.storedDisplaySettings = updateBlock([self displaySettings]);
    [self saveDisplaySettings];
}

- (void)loadDisplaySettings
{
    self.storedDisplaySettings = [NSKeyedUnarchiver unarchiveObjectWithFile:[SettingsStore pathToStore]];
}

- (void)saveDisplaySettings
{
    [NSKeyedArchiver archiveRootObject:self.storedDisplaySettings toFile:[SettingsStore pathToStore]];
}

@end

File: CommonCode/SettingsStore.swift
//
//  SettingsStore.swift
//  SpectrogeddonOSX
//
//  Created by Tom York on 26/05/2019.
//  Copyright © 2019 Spectrogeddon. All rights reserved.
//

import Foundation

class SettingsStore {
    
    private(set) var displaySettings: DisplaySettings
    
    private let storeURL: URL
    
    init(storeURL: URL = defaultStoreURL) {
        self.storeURL = storeURL
        let stored = loadSettings(at: storeURL)
        self.displaySettings = stored ?? DisplaySettings()
    }

    func applyUpdate(_ update: (DisplaySettings) -> DisplaySettings) {
        displaySettings = update(displaySettings)
        save(settings: displaySettings, to: storeURL)
    }
}

private func loadSettings(at url: URL) -> DisplaySettings? {
    precondition(url.isFileURL)
    return NSKeyedUnarchiver.unarchiveObject(withFile: url.path) as? DisplaySettings
}

private func save(settings: DisplaySettings, to url: URL) {
    precondition(url.isFileURL)
    NSKeyedArchiver.archiveRootObject(settings, toFile: url.path)
}

private let defaultStoreURL: URL = {
    
    let storeName = "Spectrogeddon.settings"
    
    let documentsURL = try! FileManager.default.url(
        for: .documentDirectory,
        in: .userDomainMask,
        appropriateFor: nil,
        create: true
    )
    
    return documentsURL.appendingPathComponent(storeName)
}()
